pkg/codegen/go/gen_test.go

```diff
@@ -66,38 +62,76 @@ func TestGoPackageName(t *testing.T) {
 
 func TestGeneratePackage(t *testing.T) {
 	generatePackage := func(tool string, pkg *schema.Package, files map[string][]byte) (map[string][]byte, error) {
+
+		for f := range files {
+			t.Logf("Ignoring extraFile %s", f)
+		}
+
 		return GeneratePackage(tool, pkg)
 	}
-	test.TestSDKCodegen(t, "go", generatePackage, typeCheckGeneratedPackage)
+	test.TestSDKCodegen(t, &test.SDKCodegenOptions{
+		Language:   "go",
+		GenPackage: generatePackage,
+		Checks: map[string]test.CodegenCheck{
+			"go/compile": typeCheckGeneratedPackage,
+			"go/test":    testGeneratedPackage,
+		},
+	})
+}
+
+func inferModuleName(codeDir string) string {
+	// For example for this path:
+	//
+	// codeDir = "../internal/test/testdata/external-resource-schema/go/"
+	//
+	// We will generate "$codeDir/go.mod" using
+	// `external-resource-schema` as the module name so that it
+	// can compile independently.
+	return filepath.Base(filepath.Dir(codeDir))
 }
 
-func typeCheckGeneratedPackage(t *testing.T, pwd string) {
-	var err error
-	var ex string
-	ex, err = executable.FindExecutable("go")
+func typeCheckGeneratedPackage(t *testing.T, codeDir string) {
+	sdk, err := filepath.Abs(filepath.Join("..", "..", "..", "sdk"))
 	require.NoError(t, err)
 
-	// go SDKs live in their own folder:
-	// typecheck/go/$NAME/$FILES
-	// where FILES contains all the project files (including go.mod)
-	//
-	// This is not true when `package.language.go.rootPackageName` is set.
-	dir, err := ioutil.ReadDir(pwd)
+	goExe, err := executable.FindExecutable("go")
 	require.NoError(t, err)
-	root := pwd
-	if len(dir) == 1 {
-		require.True(t, dir[0].IsDir())
-		root = filepath.Join(pwd, dir[0].Name())
+
+	// Remove existing `go.mod` first otherise `go mod init` fails.
+	goMod := filepath.Join(codeDir, "go.mod")
+	alreadyHaveGoMod, err := test.PathExists(goMod)
+
+	require.NoError(t, err)
+	if alreadyHaveGoMod {
+		err := os.Remove(goMod)
+		require.NoError(t, err)
 	}
-	t.Logf("*** Testing go in dir: %q ***", root)
 
+	runCommand(t, "go_mod_init", codeDir, goExe, "mod", "init", inferModuleName(codeDir))
```

Comment:
Is the correct response to delete and then renew `go.mod` or is it to create only if it doesn't exist.

Reply:
This is a tricky one. It's not coming from the codegen, and in the latest deletion scheme, it'll never be coming from old runs of codegen (which is why I was worried and making sure to delete it). It may however come from go-extras mixins. If it does the user probably intends that one to be used. I can switch to creating it only if it does not exist.

Reply:
Added the change PTAL.